docs: Updated educator docs for using Open edX as an LTI tool provider (#1448)
openedx/docs.openedx.org: source/site_ops/install_configure_run_guide/configuration/lti/configure_lti.rst

   - `Consumer Key`: A unique identifier for this consumer. Generated
     automatically but you can also supply your own.

   - `Consumer Secret`: A secret used to sign LTI requests.
     Generated automatically but you can also supply your own.

   .. important::

      Leave `Instance GUID` blank. The external LMS generates and
      supplies this value on its first LTI launch.

   - `Require User Account` and `Use LTI PII`: These checkboxes
     control how Open edX associates learners with accounts. See
     :ref:`Authentication Modes` below before deciding which to enable.

#. Select :guilabel:`Save`.

#. Share the `Consumer Key` and `Consumer Secret` with the educator
   or administrator who will configure the external LMS.

.. _Authentication Modes:

*****************************
Authentication Modes
*****************************

When a learner launches LTI content for the first time, your Open edX instance
must associate them with a user account. The two checkboxes, "Require User Account"
and "Use LTI PII", on the consumer record control how this happens. You can choose
to use the "Anonymous (default)" mode or the "Auto-Create with Personal Information"

> mode.

Reviewer: ```suggestion
to use the "Anonymous (default)" mode, the "Auto-Create with Personal Information"
mode, or the "Require Existing Account" mode.
```

.. figure:: /_images/site_ops_how_tos/lti_cosumer_admin_panel.png
   :alt: Screenshot of add LTI Consumer page showing *Require User Account* and *Use LTI PII* checkboxes.
   :width: 100%

   *Require User Account* and *Use LTI PII* checkboxes on the add LTI Consumer page in Django admin panel.

Anonymous (default)
===================

To enable this mode, leave both `Require User Account` and `Use LTI PII`
unchecked.

Your Open edX instance  will automatically create an account with a
randomly generated username and email address. The learner sees no
login prompt and the content loads immediately. This account is linked
to the learner's identity in the external LMS for grade passback.

> Use this mode when you want the most seamless experience.

Reviewer: ```suggestion
Use this mode when you want the most seamless learner experience.

.. note::
  In this mode, course authors and administrators won't be able to easily identify learners in your Open edX instance, because both their usernames and emails will be randomly generated.
```

that email already exists, it is linked automatically instead of
creating a new one.

Use this mode when you need to associate LTI activity with real learner
identities or with accounts that already exist on your Open edX instance.

The external LMS must be configured to send ``lis_person_contact_email_primary``
in the launch request. This has been tested with Canvas LMS with the
privacy setting set to "Email Only" or "Public".

Require Existing Account
========================

To enable this mode, check `Require User Account`.

On the learner's first LTI launch, your Open edX instance checks two conditions:

- The learner is already signed into your Open edX instance in the same browser.
- Their account email matches ``lis_person_contact_email_primary``
  sent by the external LMS.

If either condition is not met, the learner sees an error page with a
link to the your instance's sign-in page (Authn MFE). After signing in with the matching
account, they can return to the content.

.. important::

   The external LMS must be configured to send
   ``lis_person_contact_email_primary``. Without it, this check always
   fails, regardless of whether the learner is signed in.

> This check runs only on the first launch per learner. Once their identities have been,

Reviewer: ```suggestion
This check runs only on the first launch per learner. Once their identities have been linked,
```

.. _Configuring Credentials for a Tool Consumer:

###############################################################
Configuring LTI Consumers
###############################################################

.. tags:: site operator, how-to

For each external LMS that will consume content from your Open edX
instance, you must create a Learning Tools Interoperability (LTI) consumer record in the Django admin.
Each external LMS must have its own separate record.

.. contents::
   :local:
   :depth: 1

*****************************
Create a Consumer Record
*****************************

#. Sign in to the Django administration console at
   ``https://{your_lms_domain}/admin``.

#. Under `LTI Provider`, select :guilabel:`Add` next to `LTI Consumers`.

#. Fill in the fields:

   - `Consumer Name`: A label identifying this external LMS (for

>      example, ``Canvas - Example University``).

Reviewer: Is this name arbitrary / just for the site operator's reference, or does it mean anything?

Author: Arbitrary. But i've removed it now.

   - `Consumer Key`: A unique identifier for this consumer. Generated
     automatically but you can also supply your own.

   - `Consumer Secret`: A secret used to sign LTI requests.
     Generated automatically but you can also supply your own.

   .. important::

      Leave `Instance GUID` blank. The external LMS generates and
      supplies this value on its first LTI launch.

   - `Require User Account` and `Use LTI PII`: These checkboxes
     control how Open edX associates learners with accounts. See
     :ref:`Authentication Modes` below before deciding which to enable.

#. Select :guilabel:`Save`.

#. Share the `Consumer Key` and `Consumer Secret` with the educator
   or administrator who will configure the external LMS.

.. _Authentication Modes:

*****************************
Authentication Modes
*****************************

> When a learner launches LTI content for the first time, Open edX must

Reviewer: ```suggestion
When a learner launches LTI content for the first time, your Open edX instance must
```
Not sure how best to make "Open edX" an adjective here, feel free to adjust.

Author: Done

> associate them with a user account. The two checkboxes on the consumer

Reviewer: ```suggestion
associate them with a user account. The two checkboxes, "Require User Account" and "Use LTI PII", on the consumer
```

I'm adding this clarification because the next header is "Anonymous" but that isn't actually the name of either checkbox, so it's convoluted to read currently.

Author: Good catch. Updated.

> record control how this happens.

Reviewer: ```suggestion
record control how this happens. You can choose to use the "Anonymous (default)" mode or the "Auto-Create with Personal Information" mode.
```

Again, clarifying this preamble to orient the reader to the next section.

Author: Done

Anonymous (default)
===================

> Both `Require User Account` and `Use LTI PII` are unchecked.

Reviewer: I don't know if this is the best way to present this information. I would either leave as-is (but add a screenshot), or rework maybe something like

```
.. tip::

   To set the Anonymous mode, ensure that both `Require User Account` and `Use LTI PII` are unchecked.
```

Author: I've adjusted it in a different way.

> Open edX automatically creates an account with a randomly generated

Reviewer: ```suggestion
Your Open edX instance  will automatically create an account with a randomly generated
```

Author: Done

username and email address. The learner sees no login prompt and the
content loads immediately. This account is linked to the learner's
identity in the external LMS for grade passback.

> Use this mode when learner identity on Open edX is not important and

Reviewer: What does it mean for identity to "not be important"? I'm a bit confused by this line.

Author: This just means that educators won't be able to identify learners in Open edX easily because both their usernames and emails will be randomly generated. But i see that it's confusing and probably not necessary so i've removed it.

Reviewer: I think it is important, I'm going to add this info in.

Author: I've added it in educator docs because site operators likely don't deal with individual learners.

you want the most seamless experience.

Auto-Create with Personal Information
======================================

> Check `Use LTI PII`.

Reviewer: Again I'd use a screenshot or a `.. tip::` directive

Author: Modified the text.

> Open edX uses the learner's email address and full name from the LTI

Reviewer: I'm confused a bit - is this if the learner's email/name is _already on your Open edX instance_ or is _on the external LMS_?

Also Open edX as an adjective here but I'm not sure what to suggest because I'm not clear on the context

Author: No. Learner's name and email will be included in the API call that the LMS makes to open edx instance. Open edx will use that info to create an account. I'm updated the adjective thing.

launch request (``lis_person_contact_email_primary``,
``lis_person_name_full``) to create their account. If an account with
that email already exists, it is linked automatically instead of
creating a new one.

Use this mode when you need to associate LTI activity with real learner
identities or with accounts that already exist on your Open edX instance.

The external LMS must be configured to send ``lis_person_contact_email_primary``
in the launch request. This has been tested with Canvas LMS with the
privacy setting set to "Email Only" or "Public".

Require Existing Account
========================

> Check `Require User Account`.

Reviewer: Again I'd use a screenshot or a `.. tip::` directive

Author: Modified the text.

> On the learner's first LTI launch, Open edX checks two conditions:

Reviewer: ```suggestion
On the learner's first LTI launch, your Open edX instance checks two conditions:
```

Author: Done

> - The learner is already signed into Open edX in the same browser.

Reviewer: ```suggestion
- The learner is already signed into your Open edX instance in the same browser.
```

Author: Done

> - Their Open edX account email matches ``lis_person_contact_email_primary``

Reviewer: ```suggestion
- Their account email matches ``lis_person_contact_email_primary``
```

I suggest just removing "Open edX" here, and you could probably choose to drop it in some of your other references. It starts to get confusing when you have to really explain that difference between a _specific instance_ versus the software as a whole.

Author: Done

  sent by the external LMS.

If either condition is not met, the learner sees an error page with a

> link to the Open edX sign-in page. After signing in with the matching

Reviewer: ```suggestion
link to your instance's sign-in page (Authn MFE). After signing in with the matching
```
Since this is a site ops doc referencing the MFE is okay to do

Author: Done

account, they can return to the content.

.. important::

   The external LMS must be configured to send
   ``lis_person_contact_email_primary``. Without it, this check always
   fails, regardless of whether the learner is signed in.

This check runs only on the first launch per learner. Once Open edX

> has linked their identities, subsequent launches proceed without the

Reviewer: ```suggestion
This check runs only on the first launch per learner. Once their identities have been
linked, subsequent launches proceed without the
```

Author: Done

check.

Use this mode when learners must have pre-existing accounts on your
Open edX instance and you want activity tied to those accounts.

> If both `Require User Account` and `Use LTI PII` are checked,

Reviewer: This seems important enough to put in a callout box maybe a `.. note::`?

Author: Done

`Require User Account` takes precedence.

> Delivering Content in an iframe with Require User Account

Reviewer: A little confused on headings/terminology here. "Require User Account" is a checkbox, whereas "Require Existing Account" is a mode. Next sentence says "this mode" not "this checkbox". Further this is a sub-heading under "Require Existing Account". Should this header be "Delivering Content in an iframe with Require **Existing** Account"? (emphasis mine)

Author: I've adjusted this text to make it consistent with heading.

---------------------------------------------------------

When this mode is used and content is embedded in an iframe, browsers
may block the session cookie. Add the following settings to your Tutor

> plugin under ``openedx-lms-common-settings``:

Reviewer: I think this is slightly unclear, there isn't just one global "Tutor plugin". You need to make a new Tutor plugin to actually set these things. And in that case it can be really helpful to just provide the whole plugin, not just the settings themselves (see https://github.com/openedx/docs.openedx.org/pull/1377/changes)

Author: @feanil can you please help rewrite the plugin thing?

Reviewer 2: I added it as a suggested code change [here](https://github.com/openedx/docs.openedx.org/pull/1448/changes#r3202113886)  Let me know if that's what you were thinking or if you wanted a fully separate small tutor plugin.

Reviewer: I think that's right. I'd add a `.. _ref` link to that section, and just link that section here (so you're not duplicating information).

Author: Done. I've referenced the plugin here. Thank you both.

.. code-block:: python

    SESSION_COOKIE_SAMESITE = 'None'
    SESSION_COOKIE_SECURE = True
    SESSION_COOKIE_SAMESITE_FORCE_ALL = True
    CSRF_COOKIE_SECURE = True
    CSRF_COOKIE_SAMESITE = 'None'
    X_FRAME_OPTIONS = "ALLOW-FROM <your-lti-consumer-domain>"

After adding these settings, run:

.. code-block:: bash

    tutor config save

>     tutor local restart lms

Reviewer: ```suggestion
```

We try not to include Tutor commands in case they change. I'd link the Tutor plugins page (https://docs.tutor.edly.io/tutorials/plugin.html#modifying-existing-files-with-patches) before you show the Tutor plugin code and leave it at that.

Author: I've removed the code and added link.

*****************************
Caveats
*****************************

- Enabling `Require User Account` only affects future LTI launches.
  Existing anonymous account data is not migrated.
- Disabling `Require User Account` after it has been enabled does not
  unlink accounts. If a rollback is needed, create a new consumer record

>   with the flag disabled and use the new credentials in the external LMS.

Reviewer: By "rollback" to you mean "unlink accounts"? 

This feels pretty vague, is there any more detail we could provide?

Author: This is confusing. I've removed it. It's not necessary to mention any alternates here.

*****************************
Grade Aggregation Delay
*****************************

When the external LMS links to a unit or subsection (rather than a

> single problem), Open edX aggregates grades across all problems before

Reviewer: ```suggestion
single :ref:`component <Add a Component>`), the Open edX software aggregates grades across all problems before
```

It could be useful to link to the other structural pages in this document to clarify/align on terminology (https://docs.openedx.org/en/latest/educators/how-tos/course_development/manage_components.html#add-a-component, https://docs.openedx.org/en/latest/educators/concepts/open_edx_platform/about_course_units.html#about-course-units, https://docs.openedx.org/en/latest/educators/concepts/open_edx_platform/about_course_subsections.html)

Author: I've added these links in the educator docs instead. I've added a section titled `Content You Can Deliver`.

passing them back. Individual problem components return grades
immediately. The delay applies only to unit and subsection level links.

The default aggregation interval is 15 minutes. To change it, add the
following to your Tutor plugin under ``openedx-lms-common-settings``:

.. code-block:: python

    LTI_AGGREGATE_SCORE_PASSBACK_DELAY = 15 * 60  # seconds

Replace ``15 * 60`` with your desired interval in seconds. Then apply
the change:

.. code-block:: bash

    tutor config save

>     tutor local restart lms

Reviewer: ```suggestion
Replace ``15 * 60`` with your desired interval in seconds. 
```

Author: Done

Reviewer: @ayub02 if you've addressed comments, you can just click "Resolve Conversation":

<img width="224" height="55" alt="image" src="https://github.com/user-attachments/assets/fb466241-494a-4968-af87-af7729aab29a" />

You can also, in the inline view (https://github.com/openedx/docs.openedx.org/pull/1448/changes) click on "Add suggestion to batch" (for direct changes to the code that you agree with):

<img width="223" height="130" alt="image" src="https://github.com/user-attachments/assets/89863c54-7865-4f92-a679-a59758b8378e" />

and then when you've gathered these, at the top of the page, choose "Commit Suggestions"

<img width="196" height="69" alt="image" src="https://github.com/user-attachments/assets/844f2c59-02b7-4cfe-8b13-4a9b063cab4f" />

And then those changes will be applied in a commit, and the comments will auto-resolve.

Author: Thanks. I tried commit suggestions earlier but that rendered my local copy out of sync. It was a bit of work to get them synced again. Which is why i'm making these changes locally.

.. seealso::

>    :ref:`Using Open edX as an LTI Tool Provider` (educator)

Reviewer: There's probably more that could be added here

Reviewer: For example, the other Site Operators docs that relate to this topic.

**Maintenance chart**

+--------------+-------------------------------+----------------+--------------------------------+
| Review Date  | Working Group Reviewer        |   Release      |Test situation                  |
+--------------+-------------------------------+----------------+--------------------------------+

> | 2026-05-06   | Aamir Ayub                    | Verawood       |  Pass                          |

Reviewer: Same note on not deleting history

Author: Done
